go/rep_replication.go

```diff
@@ -0,0 +1,123 @@
+package pb
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/ipfs/go-cid"
+)
+
+const header = "rtrade-replication v0"
+
+// New creates a new Replication file, with the correct header
+func New() *Replication {
+	return &Replication{
+		Header: header,
+	}
+}
+
+// Valid checks if this is an Valid Replication file, with:
+// - correct header (if strict, then no unknown extensions),
+// - at least one CID,
+// - if strict, then equal or more servers than required for replication, and
+// - replication factor greater than 1.
+// reports the first error found.
+//
+// If strict, then invalid CIDs, Servers, and unknown protocol extensions are also not allowed.
+func (r *Replication) Valid(strict bool) error {
+	if _, err := r.GetHeaderString(strict); err != nil {
+		return err
+	}
+
+	if strict && haveUnrecognized(r) {
+		return errors.New("found unrecognized data")
+	}
+
+	CIDs, err := r.GetCIDs()
+	if (strict || len(CIDs) == 0) && err != nil {
+		return err
+	}
+	as := ServerSourcesToAddrInfos(r.Servers)
+	if len(as) < int(r.GetReplicationFactor()) {
+		return fmt.Errorf("not enough servers list %v of %v", len(as), r.GetReplicationFactor())
+	}
+	for _, a := range as {
+		if _, err := a.GetID(); err != nil {
+			if strict {
+				return err
+			}
+			continue
+		}
+		if strict {
+			if _, err := a.GetMultiAddrs(); err != nil {
+				return err
+			}
+		}
+	}
+	if r.GetReplicationFactor() < 2 {
+		return fmt.Errorf("replication factor of %v is too low", r.GetReplicationFactor())
+	}
+	return nil
+}
+
+// GetHeaderString returns the full header string.
+// Returns an error if header is not valid.
+//
+// If strict, then unknown extensions are not allowed.
+func (r *Replication) GetHeaderString(strict bool) (string, error) {
+	if strict {
+		if r.Header != header {
+			return "", fmt.Errorf(`header "%s" does not match "%s"`, r.Header, header)
+		}
+	} else {
+		if !strings.HasPrefix(r.Header, header) {
+			return "", fmt.Errorf(`header "%s" does not start with "%s"`, r.Header, header)
+		}
+	}
```

Comment:
can be simplified to:

```
if strict && r.Header != header {
    return "", fmt.Errorf(`header "%s" does not match "%s"`, r.Header, header)
} else if !strings.HasPrefix(r.Header, header) {
    return "", fmt.Errorf(`header "%s" does not start with "%s"`, r.Header, header)
}

Reply:
The else if will be combined. The if if can waste some cpu cycles if combined.

Reply:
smart